domain/src/main/java/org/oppia/android/domain/topic/TopicController.kt

```diff
@@ -252,17 +252,23 @@ class TopicController @Inject constructor(
       return true
     }
 
-    if (topic.storyCount != topicProgress.storyProgressCount &&
-      topicProgress.storyProgressMap.isNotEmpty()
-    ) {
-      return true
+    topic.storyList.forEach { storySummary ->
```

Comment:
Hmm. Can this entire method (``checkIfTopicIsOngoing``) be implemented with just a single adaption of the new code below:

```kotlin
private fun checkIfTopicIsOngoing(topic: Topic, topicProgress: TopicProgress): Boolean {
  // If there's at least one story with progress and not yet completed, then the topic is considered
  // ongoing.
  return topic.storyList.any { storySummary ->
    topicProgress.storyProgressMap[storySummary.storyId]?.let { storyProgress ->
      storySummary.isOngoing(storyProgress)
    } ?: false
  }
}

private fun StorySummary.isOngoing(storyProgress: StoryProgress): Boolean {
  val playStates = storyProgress.chapterProgressMap.values.map(ChapterProgress::chapterPlayState)
  return ChapterPlayState.STARTED_NOT_COMPLETED in playStates
}
```

It seems to me that all we need to verify is at least one story is currently ongoing. Would that suffice?

Reply:
> Hmm. Can this entire method (`checkIfTopicIsOngoing`) be implemented with just a single adaption of the new code below:
> 
> ```kotlin
> private fun checkIfTopicIsOngoing(topic: Topic, topicProgress: TopicProgress): Boolean {
>   // If there's at least one story with progress and not yet completed, then the topic is considered
>   // ongoing.
>   return topic.storyList.any { storySummary ->
>     topicProgress.storyProgressMap[storySummary.storyId]?.let { storyProgress ->
>       storySummary.isOngoing(storyProgress)
>     } ?: false
>   }
> }
> 
> private fun StorySummary.isOngoing(storyProgress: StoryProgress): Boolean {
>   val playStates = storyProgress.chapterProgressMap.values.map(ChapterProgress::chapterPlayState)
>   return ChapterPlayState.STARTED_NOT_COMPLETED in playStates
> }
> ```
> 
> It seems to me that all we need to verify is at least one story is currently ongoing. Would that suffice? I'm looking to see if the overall approach can be simplified given the invariant itself is now simpler than before.

Thanks for the review. 
Yes, verifying that at least one story of `storyProgressMap` is ongoing is totally sufficient to consider that the `Topic` is in the state `In Progress`. But I think that the method `StorySummary.isOngoing` which decides that a story is Ongoing or not lacks some necessary checks.
Let us consider a Topic that has only one story which consists of 3 chapters in total. Now, a learner initiates 2 chapters and finished both of them, but he never started the 3rd chapter, now as per the logic of `StorySummary.isOngoing` the story does not have any chapter in the `ChapterPlayState.STARTED_NOT_COMPLETED` because the learner completed the two chapters which he started. And therefore the function will return that this story is not ongoing. But we know that it is still ongoing because all the chapters of the story are not completed yet, so we can't consider the story as complete. Hence, this function does not cover all the conditions for ensuring a story to be ongoing. It's missing the necessary check that, the last chapter of the story must also be present in the chapterProgressMap.

Also, I am thinking to modify the `checkIfTopicIsOngoing` in this way:

```kotlin
  private fun checkIfTopicIsOngoing(topic: Topic, topicProgress: TopicProgress): Boolean {
    topic.storyList.forEach { storySummary ->
      if (topicProgress.storyProgressMap.containsKey(storySummary.storyId)) {
        val storyProgress = topicProgress.storyProgressMap[storySummary.storyId]
        val lastChapterSummary = storySummary.chapterList.last()
        if (!storyProgress!!.chapterProgressMap.containsKey(lastChapterSummary.explorationId)) {
          return true
        }
      }
    }

    topic.storyList.forEach { storySummary ->
      if (topicProgress.storyProgressMap.containsKey(storySummary.storyId)) {
        val storyProgress = topicProgress.storyProgressMap[storySummary.storyId]
        val lastChapterSummary = storySummary.chapterList.last()
        val lastChapterProgress =
          storyProgress!!.chapterProgressMap[lastChapterSummary.explorationId]
        if (lastChapterProgress!!.chapterPlayState != ChapterPlayState.COMPLETED) {
          return true
        }
      }
    }
    return false
  }
```
For checking that, if any story of the `storyProgressMap` is completed or not, we just need to ensure that, the last chapter is present in its `chapterProgressMap` and also it must be in the `ChapterPlayState.COMPLETED` chapterPlayState. These two checks are sufficient for considering a story to be completed. And if all the stories of the `storyProgressMap` satisfy these two conditions, then we can consider the `Topic` to be `Completed`.
@BenHenning @rt4914 Please share your thoughts on this, and please correct me if I am wrong.

Reply:
Ah, this is what comes from me trying to be clever. :) My original version of isOngoing was checking whether the last chapter was completed, instead, similar to the previous logic. I think we can update it to be something like:

```kotlin
private fun StorySummary.isOngoing(storyProgress: StoryProgress): Boolean {
  val playStates = storyProgress.chapterProgressMap.values.map(ChapterProgress::chapterPlayState)
  val firstChapterState = playStates.first()
  val lastChapterState = playStates.last()
  return firstChapterState != ChapterPlayState.NOT_STARTED && lastChapterState != ChapterPlayState.COMPLETED
}
```

This will return false if the first top isn't started, or the last topic is completed. We don't have to check any of the other chapters since the app guarantees they must be played in order.

Would this work?

Reply:
Thanks for the review. I would like to point out two things here:
1. `storyProgressMap` includes only those stories, which have been started, and if a story is started, it implies that the `firstChapterState` != `ChapterPlayState.NOT_STARTED` will always be true. Hence, we don't have to check for this explicitly.
2. For checking the state of last chapter of the story, we first have to check that, the last chapter is present in the `chapterProgressMap` or not. In your current function, the last chapter here is not the actual last chapter of the story, it is the last chapter which the learner started and which was then added to the `chapterProgressMap`. 
So, if a story consists of 3 chapters, then only completing the first chapter, results into declaring the story as `Complete` because first chapter is also the last chapter as per the logic of this function. 
Hence, we just have to do two things for checking a story to be `onGoing`. First is checking that the last chapter Is present in the `chapterProgressMap` and then check its `ChapterPlayState`. These two are the necessary and sufficient conditions.
```kotlin
  private fun checkIfTopicIsOngoing(topic: Topic, topicProgress: TopicProgress): Boolean {
    topic.storyList.forEach { storySummary ->
      if (topicProgress.storyProgressMap.containsKey(storySummary.storyId)) {
        val storyProgress = topicProgress.storyProgressMap[storySummary.storyId]
        val lastChapterSummary = storySummary.chapterList.last()
        if (!storyProgress!!.chapterProgressMap.containsKey(lastChapterSummary.explorationId)) {
          return true
        }
      }
    }

    topic.storyList.forEach { storySummary ->
      if (topicProgress.storyProgressMap.containsKey(storySummary.storyId)) {
        val storyProgress = topicProgress.storyProgressMap[storySummary.storyId]
        val lastChapterSummary = storySummary.chapterList.last()
        val lastChapterProgress =
          storyProgress!!.chapterProgressMap[lastChapterSummary.explorationId]
        if (lastChapterProgress!!.chapterPlayState != ChapterPlayState.COMPLETED) {
          return true
        }
      }
    }
    return false
  }
```

Reply:
It seems that we can adapt that, then. My goal here is to simplify the current approach since it seems more complicated than what we need. Part of my suggestion above was meant to be more of a "let's see if we can simplify it, here's one option," but as you pointed out I hadn't thought through it carefully enough. What about the following:

```kotlin
/** Return whether the current [StorySummary] can be considered "ongoing" given the specified [StoryProgress] (that is, at least one chapter has started and the final chapter isn't yet completed). */
private fun StorySummary.isOngoing(storyProgress: StoryProgress): Boolean {
  val firstChapterState = storyProgress.getChapterPlayState(chapterList.first().explorationId)
  val lastChapterState = storyProgress.getChapterPlayState(chapterList.last().explorationId)
  return firstChapterState != ChapterPlayState.NOT_STARTED && lastChapterState != ChapterPlayState.COMPLETED
}

/** Returns the [ChapterPlayState] of this progress for the specified exploration, or [ChapterPlayState.NOT_STARTED] if the exploration hasn't even been attempted yet. */
private fun StoryProgress.getChapterPlayState(explorationId: String): ChapterPlayState {
  return chapterProgressMap[explorationId]?.chapterPlayState ?: ChapterPlayState.NOT_STARTED
}
```

(code may need line wrapping)

My last iteration missed the fact that playStates needs to be a map from chapters to chapter progress, but the above solves the problem in an easier way.

Reply:
This works fine! I have updated the function now.